src/analysis/inner_product_space/spectrum.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2021 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import analysis.inner_product_space.rayleigh
+import analysis.inner_product_space.pi_L2
+
+/-! # Spectral theory of self-adjoint operators
+
+This file covers the spectral theory of self-adjoint operators on an inner product space.
+
+The first part of the file covers general properties, true without any condition on boundedness or
+compactness of the operator or finite-dimensionality of the underlying space, notably:
+* `is_self_adjoint.conj_eigenvalue_eq_self`: the eigenvalues are real
+* `is_self_adjoint.orthogonal_family_eigenspaces`: the eigenspaces are orthogonal
+* `is_self_adjoint.orthogonal_supr_eigenspaces`: the restriction of the operator to the mutual
+  orthogonal complement of the eigenspaces has, itself, no eigenvectors
+
+The second part of the file covers properties of self-adjoint operators in finite dimension.  The
+definition `is_self_adjoint.diagonalization` provides a linear isometry equivalence from a space
+`E` to the direct sum of the eigenspaces of a self-adjoint operator `T` on `E`.  The theorem
+`is_self_adjoint.diagonalization_apply_self_apply` states that, when `T` is transferred via this
+equivalence to an operator on the direct sum, it acts diagonally.  This is the *diagonalization
+theorem* for self-adjoint operators on finite-dimensional inner product spaces.
+
+## TODO
+
+Spectral theory for compact self-adjoint operators, bounded self-adjoint operators.
+
+## Tags
+
+self-adjoint operator, spectral theorem, diagonalization theorem
+
+-/
+
+variables {𝕜 : Type*} [is_R_or_C 𝕜]
+variables {E : Type*} [inner_product_space 𝕜 E]
+
+local notation `⟪`x`, `y`⟫` := @inner 𝕜 E _ x y
+
+local attribute [instance] fact_one_le_two_real
+
+open_locale classical big_operators complex_conjugate
+open module.End
+
+namespace is_self_adjoint
+
+variables {T : E →ₗ[𝕜] E} (hT : is_self_adjoint T)
+include hT
+
+/-- A self-adjoint operator preserves orthogonal complements of its eigenspaces. -/
+lemma invariant_orthogonal_eigenspace (μ : 𝕜) : ∀ v ∈ (eigenspace T μ)ᗮ, T v ∈ (eigenspace T μ)ᗮ :=
+begin
+  intros v hv w hw,
```

Comment:
```suggestion
lemma invariant_orthogonal_eigenspace (μ : 𝕜) (v : E) (hv : v ∈ (eigenspace T μ)ᗮ) :
  T v ∈ (eigenspace T μ)ᗮ :=
begin
  intros w hw,
```
which makes it clear that `v` is explicit. Also fine with me if you think it should be implicit.

Reply:
This needs to match the hypothesis on `linear_map.restrict`, so it should be explicit.